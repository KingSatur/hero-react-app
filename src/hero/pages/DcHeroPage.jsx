import { HeroList } from "../components/HeroList";

export const DcHeroPage = () => {
  return (
    <div className="pt-5">
      <HeroList publisher="DC Comics" />
    </div>
  );
};
