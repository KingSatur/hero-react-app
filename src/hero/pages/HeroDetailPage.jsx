import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroById } from "../helpers/getHeroById";

export const HeroDetailPage = () => {
  const { id: heroId = "" } = useParams();

  const navigate = useNavigate();

  const heroData = useMemo(() => getHeroById(heroId), [heroId]);

  const handleBack = () => {
    navigate("/dc");
  };

  if (!heroData) {
    return <Navigate to={"/marvel"} />;
  }

  return (
    <div className="row mt-5 animate__animated animate__fadeInLeft">
      <div className="col-4">
        <img
          src={`/assets/heroes/${heroId}.jpg`}
          className="img-thumbnail"
          alt={heroData?.superhero}
        />
      </div>
      <div className="col-8">
        <h3>{heroData?.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego:</b> {heroData?.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher:</b> {heroData?.publisher}
          </li>
          <li className="list-group-item">
            <b>First appareance:</b> {heroData?.first_appearance}
          </li>
        </ul>
        <h5 className="mt-5">Characters</h5>
        <p>{heroData?.characters}</p>

        <button className="btn btn-outline-info" onClick={handleBack}>
          Back
        </button>
      </div>
    </div>
  );
};
