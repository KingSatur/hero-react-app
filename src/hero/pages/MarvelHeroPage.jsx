import { HeroList } from "../components/HeroList";

export const MarvelHeroPage = () => {
  return (
    <div aria-label="marvel-page" className="pt-5">
      <HeroList publisher="Marvel Comics" />
    </div>
  );
};
