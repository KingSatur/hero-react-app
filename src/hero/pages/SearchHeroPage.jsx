import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../hook/useForm";
import queryString from "query-string";
import { getHerosBySearch } from "../helpers";
import { HeroCard } from "../components/HeroCard";

export const SearchHeroPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);

  const { searchText, onInputChange, onResetForm } = useForm({
    searchText: q,
  });
  const heros = getHerosBySearch(String(q));

  const showSearch = q?.length === 0;
  const showError = q && q?.length > 0 && !heros.length;

  const onSearchSubmit = (e) => {
    e?.preventDefault();

    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input
              type="text"
              name="searchText"
              placeholder="Search hero"
              autoComplete="off"
              className="form-control"
              onChange={onInputChange}
              value={searchText}
            />
            <button className="btn btn-outline-primary mt-1">Search</button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />

          <div
            aria-label="no-heros"
            className="alert alert-primary animate__animated animate__fadeInRight"
            style={{ display: showSearch ? "" : "none" }}
          >
            Search a hero
          </div>

          <div
            aria-label="no-heros-found"
            className="alert alert-danger animate__animated animate__fadeInRight"
            style={{ display: showError ? "" : "none" }}
          >
            There is not results with {q}
          </div>

          {heros?.map((hero) => (
            <HeroCard key={hero?.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
