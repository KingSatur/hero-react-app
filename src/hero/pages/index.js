export * from "./DcHeroPage";
export * from "./MarvelHeroPage";
