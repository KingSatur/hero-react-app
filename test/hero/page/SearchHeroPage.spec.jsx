import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchHeroPage } from "../../../src/hero/pages/SearchHeroPage";

const mockUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockUseNavigate,
}));

describe("SearchHeroPage tests", () => {
  fit("Should show with default values", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchHeroPage />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
    screen.debug();
  });

  it("Should should batman and input with value", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchHeroPage />
      </MemoryRouter>
    );

    const inputElement = screen.getByRole("textbox");
    const imgElement = screen.getByRole("img");
    const noHeroElement = screen.getByLabelText("no-heros");

    expect(noHeroElement.style?.display).toBe("none");
    console.log(noHeroElement.style);
    expect(inputElement.value).toBe("batman");
    expect(imgElement.src).toContain("/assets/heroes/dc-batman.jpg");
    screen.debug();
  });

  it("Should show error if hero does not exists", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/search?q=Ball12132"]}>
        <SearchHeroPage />
      </MemoryRouter>
    );

    const noHeroElement = screen.getByLabelText("no-heros-found");

    expect(noHeroElement.style?.display).not.toBe("none");
    expect(noHeroElement.textContent).toBe(
      "There is not results with Ball12132"
    );
  });

  it("Should navigate to the same page with different query", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=hola"]}>
        <SearchHeroPage />
      </MemoryRouter>
    );

    const inputElement = screen.getByRole("textbox");

    fireEvent.input(inputElement, {
      target: { name: "searchText", value: "goku" },
    });

    fireEvent.submit(inputElement);

    expect(mockUseNavigate).toHaveBeenCalledWith(`?q=goku`);
  });
});
