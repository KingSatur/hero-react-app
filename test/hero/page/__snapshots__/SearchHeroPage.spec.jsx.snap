// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchHeroPage tests Should show with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="no-heros"
        class="alert alert-primary animate__animated animate__fadeInRight"
      >
        Search a hero
      </div>
      <div
        aria-label="no-heros-found"
        class="alert alert-danger animate__animated animate__fadeInRight"
        style="display: none;"
      >
        There is not results with 
      </div>
    </div>
  </div>
</div>
`;
